target/linux/ipq40xx/image/generic.mk

```diff
@@ -753,6 +753,32 @@ define Device/openmesh_a62
 endef
 TARGET_DEVICES += openmesh_a62
 
+define Device/p2w_r619ac
+	$(call Device/FitzImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := P&W
+	DEVICE_MODEL := R619AC
+	SOC := qcom-ipq4019
+	DEVICE_DTS_CONFIG := config@10
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES := ipq-wifi-p2w_r619ac
+endef
+
+define Device/p2w_r619ac-64m
+	$(call Device/p2w_r619ac)
+	DEVICE_VARIANT := 64M NAND
+	IMAGES += nand-factory.bin
+	IMAGE/nand-factory.bin := append-ubi | qsdk-ipq-factory-nand
+endef
+TARGET_DEVICES += p2w_r619ac-64m
+
+define Device/p2w_r619ac-128m
+	$(call Device/p2w_r619ac)
+	DEVICE_VARIANT := 128M NAND
```

Comment:
With your comment from the commit message in mind:

> About the 128MiB variant: The stock bootloader sets size of nand to 64MiB.
But most of this devices have 128MiB nand. If you want to use all 128MiB,
you need to modify the `MIBIB` data of bootloader.

Hmm, can you provide instructions how to modifiy the MIBIB?

Can the bootloader read a kernel from ubi, if it has blocks placed at and over the 64MiB mark?
(Reason that this can happen is when UBI has to work around a bad block and remaps it "somewhere else",
so having something that just needs to be copy&pasted into a shell as part of the install instructions
would be helpful)

Reply:
It's provided by @ygleg [here](https://github.com/openwrt/openwrt/pull/3691#issuecomment-818770060). I have not tested this because I don't have tools to flash the bootloader chip. Now I'm using a third-party bootloader.

Reply:
I have tested the MIBIB file attached above, which works as intended.
It seems that NOR flash MTDs can not be recognized by `mtd` utility, making the flash procedure only viable (without an external SPI flash programmer) by erasing the full NAND, which will make stock u-boot fallback to a working shell. After that, the partition could be updated by using integrated TFTP and mtd command.
There are also 3rd-party bootloaders which support the new (same? since the only partition on NAND is the rootfs ubi) 128MB partition scheme.

Reply:
Yea, I this would be something for the wiki. For now, could you add a link to the github in the commit message? This is to point people into the right direction and to that helpful comment.

Reply:
You may also short any two of the data pins together on the NAND with a tweezer (pinout can be found in the ONFI specification or the datasheet, it's standard), that will cause u-boot to fail while loading the kernel and drop into a shell.